src/backend/puppet/mod.rs

```diff
@@ -0,0 +1,165 @@
+use backend::puppet::observed::ObservedChar;
+use theme::Effect;
+use theme::ColorPair;
+use std::cell::Ref;
+
+///
+pub struct Backend {
+    size : Vec2,
+    inner_sender: Sender<Option<Event>>,
+    inner_receiver: Receiver<Option<Event>>,
+    frames : LinkedList<ObservedFrame>,
+    current_frame : RefCell<ObservedFrame>,
+    current_color : RefCell<Rc<theme::ColorPair>>,
+    current_effect : RefCell<Rc<EnumSet<theme::Effect>>>,
+    current_cursor : Vec2,
+}
+
+impl Backend {
+    ///
+    pub fn init() -> Box<backend::Backend>
+        where
+            Self: Sized,
+    {
+        let (inner_sender, inner_receiver) = crossbeam_channel::bounded(1);
+        let size = Vec2::new(100, 100); //TODO something less random
+        let color_pair = theme::ColorPair {
+            front: theme::Color::TerminalDefault,
+            back : theme::Color::Dark(theme::BaseColor::Black)
+        };
+
+        Box::new(Backend {
+            size,
+            inner_sender,
+            inner_receiver,
+            frames : LinkedList::new(),
+            current_frame : RefCell::new(ObservedFrame::new(size)),
+            current_color : RefCell::new(Rc::new(color_pair)),
+            current_effect : RefCell::new(Rc::new(EnumSet::new())),
+            current_cursor: Vec2::new(0,0)
+        })
+    }
+
+    ///
+    pub fn current_frame(&self) -> Ref<ObservedFrame> {
+        self.current_frame.borrow()
+    }
+
+    ///
+    pub fn last_frame(&self) -> Option<&ObservedFrame> {
+        self.frames.front()
+    }
+}
+
+impl backend::Backend for Backend {
+    fn finish(&mut self) {
+        self.current_cursor = Vec2::new(0,0);
+        let mut finished_frame = self.current_frame.replace(ObservedFrame::new(self.size));
+        finished_frame.finish();
+        self.frames.push_front(finished_frame);
+    }
+
+    fn refresh(&mut self) {}
+
+    fn has_colors(&self) -> bool {
+        true
+    }
+
+    fn screen_size(&self) -> Vec2 {
+        self.size
+    }
+
+    fn prepare_input(&mut self, _input_request: backend::InputRequest) {
+        self.inner_sender.send(Some(Event::Exit));
+    }
+
+    fn start_input_thread(
+        &mut self, event_sink: Sender<Option<Event>>,
+        input_requests: Receiver<backend::InputRequest>,
+    ) {
+        let receiver = self.inner_receiver.clone();
+
+        thread::spawn(move || {
+            for _ in input_requests {
+                match receiver.recv() {
+                    Err(_) => return,
+                    Ok(event) => {
+                        if event_sink.send(event).is_err() {
+                            return;
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    fn print_at(&self, pos: Vec2, text: &str) {
+        for (idx, c) in text.graphemes(true).enumerate() {
+            let mut f = self.current_frame.borrow_mut();
+            let mut cpos = pos;
+            cpos.x += idx;
```

Comment:
Nitpicking: this could also be done without mutability with `let cpos = pos + (idx, 0);`